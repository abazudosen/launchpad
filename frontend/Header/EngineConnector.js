/* @flow */

import React, { Component } from 'react';
import gql from 'graphql-tag'
import type { Pad, Context } from '../types';
import HeaderButton from './HeaderButton';

type EngineConnectorProps = {|
  pad: Pad,
  engineClient: any,
  currentContext: Array<Context>,
  onContextChange: (Array<Context>) => any,
|};

type EngineConnectorState = {|
  data: {
    me: ?{ id: string },
    service: ?{ id: string, apiKeys: Array<{ token: string }> },
  },
  loading: boolean,
|};

export default class EngineConnector extends Component<
  void,
  EngineConnectorProps,
  EngineConnectorState,
> {
  subscription: any;
  state = {
    data: {
      me: null,
      service: null,
    },
    loading: true,
  };

  componentDidMount() {
    const observable = this.props.engineClient.watchQuery({
      query: gql`
        query RootQuery($serviceId: ID!) {
          me {
            id
          }
          service(id: $serviceId) {
            id
            apiKeys {
              token
            }
          }
        }
      `,
      variables: {
        serviceId: `launchpad-${this.props.pad.id}`,
      },
    });
    this.subscription = observable.subscribe({
      next: next => {
        this.setState({ data: next.data, loading: next.loading });
      },
    });
  }

  componentWillUnmount() {
    if (this.subscription) {
      this.subscription.unsubscribe();
    }
  }

  async handleEnableEngine() {
    if (this.state.data.me) {
      const result = await this.props.engineClient.mutate({
        mutation: gql`
          mutation($name: String!, $serviceId: ID!, $accountId: ID!) {
            newService(id: $serviceId, accountId: $accountId, name: $name) {
              id
              apiKeys {
                token
              }
            }
          }
        `,
        variables: {
          serviceId: `launchpad-${this.props.pad.id}`,
          name: this.props.pad.title || `launchpad-${this.props.pad.id}`,
          accountId: this.state.data.me.id,
        },
        refetchQueries: ['RootQuery'],
      });

      const newContext = this.props.currentContext.filter(
        ({ key }) => key !== 'APOLLO_ENGINE_KEY',
      );
      newContext.push({
        key: 'APOLLO_ENGINE_KEY',
        value: result.data.newService.apiKeys[0].token,
      });
      this.props.onContextChange(newContext);
    }
  }

  renderInner() {
    if (this.state.data.me == null) {
      const loginUrl = `https://engine-graphql.apollodata.com/login?cb=${document
        .location.href}`;
      return (
        <HeaderButton onClick={() => (document.location.href = loginUrl)}>
          Login to Apollo Engine
        </HeaderButton>
      );
    } else if (this.state.data.service == null) {
      return (
        <HeaderButton onClick={() => this.handleEnableEngine()}>
          Enable Apollo Engine
        </HeaderButton>
      );
    } else {
      const engineUrl = `https://engine.apollographql.com/service/${this.state
        .data.service.id}`;
      return (
        <HeaderButton onClick={() => window.open(engineUrl, '_blank')}>
          View in Apollo Engine
        </HeaderButton>
      );
    }
  }

  render() {
    if (
      this.props.pad.isDraft ||
      !this.state.data ||
      this.state.loading ||
      !this.props.pad.user ||
      this.props.pad.user.id !== this.props.user.id
    ) {
      return null;
    } else {
      return <div className="Header-EngineButtons">{this.renderInner()}</div>;
    }
  }
}
