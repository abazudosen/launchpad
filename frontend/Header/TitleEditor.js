/* @flow */

import React from 'react';
import Helmet from 'react-helmet';

import './TitleEditor.less';

type TitleEditorProps = {|
  title: ?string,
  canEdit: boolean,
  githubUsername: ?string,
  onSetTitle: (title: string) => any,
|};

export default class TitleEditor extends React.Component {
  props: TitleEditorProps;

  state: {|
    title: string,
  |};

  constructor(props: TitleEditorProps) {
    super();

    this.state = {
      title: props.title || '',
    };
  }

  handleTitleChange = (evt: SyntheticInputEvent) => {
    this.setState({
      title: evt.target.value,
    });
  };

  saveTitle = () => {
    if (this.state.title !== this.props.title) {
      // Only set title if it was actually changed
      this.props.onSetTitle(this.state.title);
    }
  };

  handleKeyPress = (evt: SyntheticInputEvent) => {
    if (evt.key === 'Enter') {
      this.saveTitle();
    }
  };

  renderTitle() {
    if (this.props.title) {
      return this.props.title;
    } else {
      return <span className="Header-Title--untitled">(Untitled)</span>;
    }
  }

  renderUser() {
    if (this.props.githubUsername) {
      return (
        <span>
          by{' '}
          <a
            href={`https://github.com/${this.props.githubUsername}`}
            target="_blank"
          >
            {this.props.githubUsername}
          </a>
        </span>
      );
    } else {
      return null;
    }
  }

  render() {
    if (this.props.canEdit) {
      return (
        <div className="Title-Editor">
          <Helmet>
            <meta
              name="twitter:title"
              content={
                this.state.title === '' ? 'Untitled Pad' : this.state.title
              }
            />
            <meta
              property="og:title"
              content={
                this.state.title === '' ? 'Untitled Pad' : this.state.title
              }
            />
          </Helmet>
          <div
            className="edit-title-button"
            aria-label="Edit Title and Description"
          >
            <span key="icon-edit" className="icon-edit" />
          </div>

          <input
            type="text"
            className="TitleInput"
            placeholder="Untitled"
            value={this.state.title}
            onChange={this.handleTitleChange}
            onBlur={this.saveTitle}
            onKeyPress={this.handleKeyPress}
          />
        </div>
      );
    } else {
      return (
        <div className="Title">
          <div>
            {this.renderTitle()} {this.renderUser()}
          </div>
        </div>
      );
    }
  }
}
