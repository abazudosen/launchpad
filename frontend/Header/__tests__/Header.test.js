/* @flow */

import React from 'react';
import Header from '../Header.js';
import ReactShallowRenderer from 'react-test-renderer/shallow';

const renderer = new ReactShallowRenderer();

const testPad = {
  id: 'fakeid1',
  title: 'Fake 1',
  description: 'Description 1',
  code: 'code',
  deployedCode: 'deployed code',
  url: 'http://example.com',
  user: {
    id: 'test-ids',
    githubUsername: 'example',
    pads: [],
  },
  context: [
    {
      key: 'secret',
      value: 'secret',
    },
  ],
  dependencies: [
    {
      name: 'graphql-js',
      version: '0.8.9',
    },
  ],
  defaultQuery: '{ hello }',
  defaultVariables: null,
  token: 'test-token-2',
  draft: null,
};

const testUser = testPad.user;

function renderHeader({ user, pad, isDraft, isDeployed, isDeploying }) {
  return (
    <Header
      user={user}
      pad={{
        ...pad,
        isDraft,
        isDeployed,
      }}
      engineClient={{}}
      currentContext={[]}
      onContextChange={() => {}}
      isDeploying={isDeploying}
      onDeploy={() => {}}
      onReset={() => {}}
      onFork={() => {}}
      onDownload={() => {}}
      onLogin={() => {}}
      onLogout={() => {}}
      onSetTitle={() => {}}
      onSetDescription={() => {}}
    />
  );
}

test('renders anon existing', () => {
  const component = renderer.render(
    renderHeader({
      user: null,
      pad: testPad,
      isDraft: false,
      isDeployed: true,
      isDeploying: false,
    }),
  );
  expect(component).toMatchSnapshot();
});

test('renders anon new', () => {
  const component = renderer.render(
    renderHeader({
      user: null,
      pad: {
        ...testPad,
        user: null,
      },
      isDraft: false,
      isDeployed: false,
      isDeploying: false,
    }),
  );
  expect(component).toMatchSnapshot();
});

test('renders own', () => {
  const component = renderer.render(
    renderHeader({
      user: testUser,
      pad: testPad,
      isDraft: false,
      isDeployed: true,
      isDeploying: false,
    }),
  );
  expect(component).toMatchSnapshot();
});

test('renders deploying', () => {
  const component = renderer.render(
    renderHeader({
      user: testUser,
      pad: testPad,
      isDraft: true,
      isDeployed: true,
      isDeploying: true,
    }),
  );
  expect(component).toMatchSnapshot();
});

test('renders undeployed', () => {
  const component = renderer.render(
    renderHeader({
      user: testUser,
      pad: testPad,
      isDraft: false,
      isDeployed: false,
      isDeploying: false,
    }),
  );
  expect(component).toMatchSnapshot();
});

test('renders draft', () => {
  const component = renderer.render(
    renderHeader({
      user: testUser,
      pad: testPad,
      isDraft: true,
      isDeployed: true,
      isDeploying: false,
    }),
  );
  expect(component).toMatchSnapshot();
});

test('render others', () => {
  const component = renderer.render(
    renderHeader({
      user: testUser,
      pad: {
        ...testPad,
        user: {
          id: 'some-new-id',
          githubUsername: 'some-new-username',
          pads: [],
        },
      },
      isDraft: false,
      isDeployed: true,
      isDeploying: false,
    }),
  );
  expect(component).toMatchSnapshot();
});
