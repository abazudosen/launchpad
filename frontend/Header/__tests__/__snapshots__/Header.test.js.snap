// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Heaaer render others 1`] = `
<div
  className="Header"
>
  <div
    className="Header-Left"
  >
    <HeaderLogo />
    <div
      className="Management-Buttons"
    >
      <HeaderButton
        onClick={[Function]}
      >
        New
      </HeaderButton>
      <HeaderButton
        disabled={undefined}
        onClick={undefined}
        tooltip="You can't save pads you don't own"
      >
        Save
      </HeaderButton>
      <HeaderButton
        disabled={false}
        onClick={[Function]}
        tooltip={undefined}
      >
        Download
      </HeaderButton>
      <HeaderButton
        disabled={false}
        onClick={[Function]}
        tooltip={undefined}
      >
        Fork
      </HeaderButton>
    </div>
    <TitleEditor
      canEdit={false}
      githubUsername="some-new-username"
      onSetTitle={[Function]}
      title="Fake 1"
    />
    <EngineConnector
      currentContext={Array []}
      engineClient={Object {}}
      onContextChange={[Function]}
      pad={
        Object {
          "code": "code",
          "context": Array [
            Object {
              "key": "secret",
              "value": "secret",
            },
          ],
          "defaultQuery": "{ hello }",
          "defaultVariables": null,
          "dependencies": Array [
            Object {
              "name": "graphql-js",
              "version": "0.8.9",
            },
          ],
          "deployedCode": "deployed code",
          "description": "Description 1",
          "draft": null,
          "id": "fakeid1",
          "isDeployed": true,
          "isDraft": false,
          "title": "Fake 1",
          "token": "test-token-2",
          "url": "http://example.com",
          "user": Object {
            "githubUsername": "some-new-username",
            "id": "some-new-id",
            "pads": Array [],
          },
        }
      }
      user={
        Object {
          "githubUsername": "example",
          "id": "test-ids",
          "pads": Array [],
        }
      }
    />
  </div>
  <div
    className="Header-Gap"
  />
  <div
    className="Header-Right"
  >
    <UserBar
      onLogin={[Function]}
      onLogout={[Function]}
      user={
        Object {
          "githubUsername": "example",
          "id": "test-ids",
          "pads": Array [],
        }
      }
    />
  </div>
</div>
`;

exports[`Main Heaaer renders anon existing 1`] = `
<div
  className="Header"
>
  <div
    className="Header-Left"
  >
    <HeaderLogo />
    <div
      className="Management-Buttons"
    >
      <HeaderButton
        onClick={[Function]}
      >
        New
      </HeaderButton>
      <HeaderButton
        disabled={true}
        onClick={undefined}
        tooltip="You need to login to save"
      >
        Save
      </HeaderButton>
      <HeaderButton
        disabled={false}
        onClick={[Function]}
        tooltip={undefined}
      >
        Download
      </HeaderButton>
      <HeaderButton
        disabled={true}
        onClick={undefined}
        tooltip="You need to login to fork"
      >
        Fork
      </HeaderButton>
    </div>
    <TitleEditor
      canEdit={false}
      githubUsername="example"
      onSetTitle={[Function]}
      title="Fake 1"
    />
    <EngineConnector
      currentContext={Array []}
      engineClient={Object {}}
      onContextChange={[Function]}
      pad={
        Object {
          "code": "code",
          "context": Array [
            Object {
              "key": "secret",
              "value": "secret",
            },
          ],
          "defaultQuery": "{ hello }",
          "defaultVariables": null,
          "dependencies": Array [
            Object {
              "name": "graphql-js",
              "version": "0.8.9",
            },
          ],
          "deployedCode": "deployed code",
          "description": "Description 1",
          "draft": null,
          "id": "fakeid1",
          "isDeployed": true,
          "isDraft": false,
          "title": "Fake 1",
          "token": "test-token-2",
          "url": "http://example.com",
          "user": Object {
            "githubUsername": "example",
            "id": "test-ids",
            "pads": Array [],
          },
        }
      }
      user={null}
    />
  </div>
  <div
    className="Header-Gap"
  />
  <div
    className="Header-Right"
  >
    <UserBar
      onLogin={[Function]}
      onLogout={[Function]}
      user={null}
    />
  </div>
</div>
`;

exports[`Main Heaaer renders anon new 1`] = `
<div
  className="Header"
>
  <div
    className="Header-Left"
  >
    <HeaderLogo />
    <div
      className="Management-Buttons"
    >
      <HeaderButton
        onClick={[Function]}
      >
        New
      </HeaderButton>
      <HeaderButton
        disabled={true}
        onClick={undefined}
        tooltip="You need to login to save"
      >
        Save
      </HeaderButton>
      <HeaderButton
        disabled={true}
        onClick={undefined}
        tooltip="Save this pad to download it"
      >
        Download
      </HeaderButton>
      <HeaderButton
        disabled={true}
        onClick={undefined}
        tooltip="You need to login to fork"
      >
        Fork
      </HeaderButton>
    </div>
    <TitleEditor
      canEdit={true}
      githubUsername={null}
      onSetTitle={[Function]}
      title="Fake 1"
    />
    <EngineConnector
      currentContext={Array []}
      engineClient={Object {}}
      onContextChange={[Function]}
      pad={
        Object {
          "code": "code",
          "context": Array [
            Object {
              "key": "secret",
              "value": "secret",
            },
          ],
          "defaultQuery": "{ hello }",
          "defaultVariables": null,
          "dependencies": Array [
            Object {
              "name": "graphql-js",
              "version": "0.8.9",
            },
          ],
          "deployedCode": "deployed code",
          "description": "Description 1",
          "draft": null,
          "id": "fakeid1",
          "isDeployed": false,
          "isDraft": false,
          "title": "Fake 1",
          "token": "test-token-2",
          "url": "http://example.com",
          "user": null,
        }
      }
      user={null}
    />
  </div>
  <div
    className="Header-Gap"
  />
  <div
    className="Header-Right"
  >
    <UserBar
      onLogin={[Function]}
      onLogout={[Function]}
      user={null}
    />
  </div>
</div>
`;

exports[`Main Heaaer renders deploying 1`] = `
<div
  className="Header"
>
  <div
    className="Header-Left"
  >
    <HeaderLogo />
    <div
      className="Management-Buttons"
    >
      <HeaderButton
        onClick={[Function]}
      >
        New
      </HeaderButton>
      <HeaderButton
        disabled={false}
        onClick={[Function]}
        tooltip={undefined}
      >
        Save
      </HeaderButton>
      <HeaderButton
        disabled={true}
        onClick={undefined}
        tooltip={undefined}
      >
        Download
      </HeaderButton>
      <HeaderButton
        disabled={true}
        onClick={undefined}
        tooltip={undefined}
      >
        Fork
      </HeaderButton>
    </div>
    <TitleEditor
      canEdit={true}
      githubUsername="example"
      onSetTitle={[Function]}
      title="Fake 1"
    />
    <EngineConnector
      currentContext={Array []}
      engineClient={Object {}}
      onContextChange={[Function]}
      pad={
        Object {
          "code": "code",
          "context": Array [
            Object {
              "key": "secret",
              "value": "secret",
            },
          ],
          "defaultQuery": "{ hello }",
          "defaultVariables": null,
          "dependencies": Array [
            Object {
              "name": "graphql-js",
              "version": "0.8.9",
            },
          ],
          "deployedCode": "deployed code",
          "description": "Description 1",
          "draft": null,
          "id": "fakeid1",
          "isDeployed": true,
          "isDraft": true,
          "title": "Fake 1",
          "token": "test-token-2",
          "url": "http://example.com",
          "user": Object {
            "githubUsername": "example",
            "id": "test-ids",
            "pads": Array [],
          },
        }
      }
      user={
        Object {
          "githubUsername": "example",
          "id": "test-ids",
          "pads": Array [],
        }
      }
    />
  </div>
  <div
    className="Header-Gap"
  />
  <div
    className="Header-Right"
  >
    <UserBar
      onLogin={[Function]}
      onLogout={[Function]}
      user={
        Object {
          "githubUsername": "example",
          "id": "test-ids",
          "pads": Array [],
        }
      }
    />
  </div>
</div>
`;

exports[`Main Heaaer renders draft 1`] = `
<div
  className="Header"
>
  <div
    className="Header-Left"
  >
    <HeaderLogo />
    <div
      className="Management-Buttons"
    >
      <HeaderButton
        onClick={[Function]}
      >
        New
      </HeaderButton>
      <HeaderButton
        disabled={false}
        onClick={[Function]}
        tooltip={undefined}
      >
        Save
      </HeaderButton>
      <HeaderButton
        disabled={true}
        onClick={[Function]}
        tooltip="Save this pad to download it"
      >
        Download
      </HeaderButton>
      <HeaderButton
        disabled={false}
        onClick={[Function]}
        tooltip={undefined}
      >
        Fork
      </HeaderButton>
    </div>
    <TitleEditor
      canEdit={true}
      githubUsername="example"
      onSetTitle={[Function]}
      title="Fake 1"
    />
    <EngineConnector
      currentContext={Array []}
      engineClient={Object {}}
      onContextChange={[Function]}
      pad={
        Object {
          "code": "code",
          "context": Array [
            Object {
              "key": "secret",
              "value": "secret",
            },
          ],
          "defaultQuery": "{ hello }",
          "defaultVariables": null,
          "dependencies": Array [
            Object {
              "name": "graphql-js",
              "version": "0.8.9",
            },
          ],
          "deployedCode": "deployed code",
          "description": "Description 1",
          "draft": null,
          "id": "fakeid1",
          "isDeployed": true,
          "isDraft": true,
          "title": "Fake 1",
          "token": "test-token-2",
          "url": "http://example.com",
          "user": Object {
            "githubUsername": "example",
            "id": "test-ids",
            "pads": Array [],
          },
        }
      }
      user={
        Object {
          "githubUsername": "example",
          "id": "test-ids",
          "pads": Array [],
        }
      }
    />
  </div>
  <div
    className="Header-Gap"
  />
  <div
    className="Header-Right"
  >
    <UserBar
      onLogin={[Function]}
      onLogout={[Function]}
      user={
        Object {
          "githubUsername": "example",
          "id": "test-ids",
          "pads": Array [],
        }
      }
    />
  </div>
</div>
`;

exports[`Main Heaaer renders own 1`] = `
<div
  className="Header"
>
  <div
    className="Header-Left"
  >
    <HeaderLogo />
    <div
      className="Management-Buttons"
    >
      <HeaderButton
        onClick={[Function]}
      >
        New
      </HeaderButton>
      <HeaderButton
        disabled={true}
        onClick={undefined}
        tooltip="Already up to date"
      >
        Save
      </HeaderButton>
      <HeaderButton
        disabled={false}
        onClick={[Function]}
        tooltip={undefined}
      >
        Download
      </HeaderButton>
      <HeaderButton
        disabled={false}
        onClick={[Function]}
        tooltip={undefined}
      >
        Fork
      </HeaderButton>
    </div>
    <TitleEditor
      canEdit={true}
      githubUsername="example"
      onSetTitle={[Function]}
      title="Fake 1"
    />
    <EngineConnector
      currentContext={Array []}
      engineClient={Object {}}
      onContextChange={[Function]}
      pad={
        Object {
          "code": "code",
          "context": Array [
            Object {
              "key": "secret",
              "value": "secret",
            },
          ],
          "defaultQuery": "{ hello }",
          "defaultVariables": null,
          "dependencies": Array [
            Object {
              "name": "graphql-js",
              "version": "0.8.9",
            },
          ],
          "deployedCode": "deployed code",
          "description": "Description 1",
          "draft": null,
          "id": "fakeid1",
          "isDeployed": true,
          "isDraft": false,
          "title": "Fake 1",
          "token": "test-token-2",
          "url": "http://example.com",
          "user": Object {
            "githubUsername": "example",
            "id": "test-ids",
            "pads": Array [],
          },
        }
      }
      user={
        Object {
          "githubUsername": "example",
          "id": "test-ids",
          "pads": Array [],
        }
      }
    />
  </div>
  <div
    className="Header-Gap"
  />
  <div
    className="Header-Right"
  >
    <UserBar
      onLogin={[Function]}
      onLogout={[Function]}
      user={
        Object {
          "githubUsername": "example",
          "id": "test-ids",
          "pads": Array [],
        }
      }
    />
  </div>
</div>
`;

exports[`Main Heaaer renders undeployed 1`] = `
<div
  className="Header"
>
  <div
    className="Header-Left"
  >
    <HeaderLogo />
    <div
      className="Management-Buttons"
    >
      <HeaderButton
        onClick={[Function]}
      >
        New
      </HeaderButton>
      <HeaderButton
        disabled={true}
        onClick={undefined}
        tooltip="Already up to date"
      >
        Save
      </HeaderButton>
      <HeaderButton
        disabled={true}
        onClick={undefined}
        tooltip="Save this pad to download it"
      >
        Download
      </HeaderButton>
      <HeaderButton
        disabled={true}
        onClick={undefined}
        tooltip="This pad is not saved yet"
      >
        Fork
      </HeaderButton>
    </div>
    <TitleEditor
      canEdit={true}
      githubUsername="example"
      onSetTitle={[Function]}
      title="Fake 1"
    />
    <EngineConnector
      currentContext={Array []}
      engineClient={Object {}}
      onContextChange={[Function]}
      pad={
        Object {
          "code": "code",
          "context": Array [
            Object {
              "key": "secret",
              "value": "secret",
            },
          ],
          "defaultQuery": "{ hello }",
          "defaultVariables": null,
          "dependencies": Array [
            Object {
              "name": "graphql-js",
              "version": "0.8.9",
            },
          ],
          "deployedCode": "deployed code",
          "description": "Description 1",
          "draft": null,
          "id": "fakeid1",
          "isDeployed": false,
          "isDraft": false,
          "title": "Fake 1",
          "token": "test-token-2",
          "url": "http://example.com",
          "user": Object {
            "githubUsername": "example",
            "id": "test-ids",
            "pads": Array [],
          },
        }
      }
      user={
        Object {
          "githubUsername": "example",
          "id": "test-ids",
          "pads": Array [],
        }
      }
    />
  </div>
  <div
    className="Header-Gap"
  />
  <div
    className="Header-Right"
  >
    <UserBar
      onLogin={[Function]}
      onLogout={[Function]}
      user={
        Object {
          "githubUsername": "example",
          "id": "test-ids",
          "pads": Array [],
        }
      }
    />
  </div>
</div>
`;
